lib/rust/lazy-reader/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+            self.buffer[self.offset - i] = self.buffer[len - i];
+        }
+        self.length = self.offset + self.reader.read(&mut self.buffer[self.offset..]);
+        self.offset = self.offset - words;
+    }
+
+    /// Is the reader empty.
+    pub fn empty(&self) -> bool {
+        self.length < self.buffer.len() && self.length <= self.offset
+    }
+
+    /// Reads the next char from input.
+    pub fn next_char(&mut self) -> Result<char,Error> {
+        if self.empty() { return Err(Error::EOF) }
+
+        if self.offset >= self.buffer.len() - D::MAX_SYMBOL_LEN {
+            self.fill();
+        }
+
+        self.character = D::decode(&self.buffer[self.offset..]);
+        self.rewinded  = false;
+        self.offset    = self.offset + self.character.size;
+
+        match self.character.char {
+            Some(char) => Ok(char),
+            None       => Err(Error::InvalidChar)
+        }
+    }
+}
+
+
+
+// =============
+// === Tests ===
+// =============
+
+#[cfg(test)]
+mod tests {
+    extern crate test;
+
+    use super::*;
+    use decoder::*;
+
+    use test::Bencher;
+
+
+
+    // ================
+    // === Repeater ===
+    // ================
+
+    /// Struct that holds state of `Reader` that repeats an input n times.
+    #[derive(Debug, Clone)]
+    struct Repeat<T> {
+        /// The input to be repeated.
+        buffer: Vec<T>,
+        /// The current offset of element currently read from buffer.
+        offset: usize,
+        /// How many more times the input should be repeated.
+        repeat: usize,
+    }
+
+    impl<T:Copy> Read for Repeat<T> {
+        type Item = T;
+
+        fn read(&mut self, mut buffer:&mut [Self::Item]) -> usize {
+            if self.repeat == 0 { return 0 }
+
+            let len  = self.buffer.len();
+            let read = buffer.len();
+
+            if read < len - self.offset {
+                buffer.copy_from_slice(&self.buffer[self.offset..self.offset + read]);
+                self.offset += read;
+                return read
+            }
+
+            buffer[..len - self.offset].copy_from_slice(&self.buffer[self.offset..]);
+            buffer = &mut buffer[len - self.offset..];
+
+            let repeat  = std::cmp::min(buffer.len() / len, self.repeat - 1);
+            self.repeat = self.repeat - repeat - 1;
+            for _ in 0..repeat {
+                buffer[..len].copy_from_slice(&self.buffer[..]);
+                buffer = &mut buffer[len..];
+            }
+
+            if self.repeat == 0 {
+                return len - self.offset + repeat * len
+            }
+            buffer.copy_from_slice(&self.buffer[..buffer.len()]);
+            self.offset = buffer.len();
+            read
+        }
+    }
+
+    /// Creates a reader that repeats an input n times.
+    fn repeat<T:Copy>(input:Vec<T>, repeat:usize) -> impl Read<Item=T> {
+        Repeat { buffer:input, repeat, offset: 0 }
+    }
+
+
+
+    // =============
+    // === Tests ===
```

Comment:
Can you benchmark memory usage without too much effort? If so, please add benchmarks that let us observe this reading in constant (low) memory.